src/daft-dsl/src/join.rs

```diff
@@ -0,0 +1,99 @@
+    right_schema: &'a SchemaRef,
+) -> impl Iterator<Item = &'a String> {
+    left_schema
+        .fields
+        .keys()
+        .filter(|name| right_schema.has_field(name))
+}
+
+/// Infer the schema of a join operation
+pub fn infer_join_schema(
+    left_schema: &SchemaRef,
+    right_schema: &SchemaRef,
+    join_type: JoinType,
+) -> DaftResult<SchemaRef> {
+    if matches!(join_type, JoinType::Anti | JoinType::Semi) {
+        Ok(left_schema.clone())
+    } else {
+        let common_cols = get_common_join_cols(left_schema, right_schema).collect::<IndexSet<_>>();
+
+        // common columns, then unique left fields, then unique right fields
+        let fields = common_cols
+            .iter()
+            .map(|name| {
+                let left_field = left_schema.get_field(name).unwrap();
+                let right_field = right_schema.get_field(name).unwrap();
+
+                Ok(match join_type {
+                    JoinType::Inner => left_field.clone(),
+                    JoinType::Left => left_field.clone(),
+                    JoinType::Right => right_field.clone(),
+                    JoinType::Outer => {
+                        let supertype = try_get_supertype(&left_field.dtype, &right_field.dtype)?;
+
+                        Field::new(*name, supertype)
+                    }
+                    JoinType::Anti | JoinType::Semi => unreachable!(),
+                })
+            })
+            .chain(
+                left_schema
+                    .fields
+                    .iter()
+                    .chain(right_schema.fields.iter())
+                    .filter_map(|(name, field)| {
+                        if common_cols.contains(name) {
+                            None
+                        } else {
+                            Some(field.clone())
+                        }
+                    })
+                    .map(Ok),
+            )
+            .collect::<DaftResult<_>>()?;
+
+        Ok(Schema::new(fields)?.into())
+    }
+}
+
+/// Casts join keys to the same types and make their names unique.
+pub fn normalize_join_keys(
+    left_on: Vec<ExprRef>,
+    right_on: Vec<ExprRef>,
+    left_schema: SchemaRef,
+    right_schema: SchemaRef,
+) -> DaftResult<(Vec<ExprRef>, Vec<ExprRef>)> {
+    let (left_on, right_on) = left_on
+        .into_iter()
+        .zip(right_on)
+        .map(|(mut l, mut r)| {
+            let l_dtype = l.to_field(&left_schema)?.dtype;
+            let r_dtype = r.to_field(&right_schema)?.dtype;
+
+            let supertype = try_get_supertype(&l_dtype, &r_dtype)?;
+
+            if l_dtype != supertype {
+                l = l.cast(&supertype);
+            }
+
+            if r_dtype != supertype {
+                r = r.cast(&supertype);
+            }
+
+            Ok((l, r))
+        })
+        .collect::<DaftResult<(Vec<_>, Vec<_>)>>()?;
+
+    let left_on = deduplicate_expr_names(&left_on);
+    let right_on = deduplicate_expr_names(&right_on);
```

Comment:
nit: can we make `deduplicate_expr_names` take in an iter so we don't have to materialize twice?

Reply:
We need to materialize here anyway to get the DaftResult out of the iterator as well as split the iterator into two vecs.